rfcs/text/0000-matrix-service.md

```diff
@@ -0,0 +1,406 @@
+| SS2               | `cell_suspension_*.json`<br />`*.isoforms.results`<br />`*.genes.results` | For SS2 pipelines, expression counts are parsed for both gene and transcript IDs from `*.genes.results` and `*.isoforms.results` files. | Since SS2 bundles conventionally contain one cell per bundle, the `cell_suspension.provenance.document_id` is used as a cell's unique identifier within the index. |
+| Optimus           | `*zarr*`<br />`empty_drops_result.csv`                       | For Optimus pipelines, expression data is parsed from the pipeline generated `zarr` according to this [specification](https://github.com/HumanCellAtlas/skylab/blob/master/docs/matrix_format_spec.md) maintained by Secondary Analysis and Matrix teams. However, a large majority of the barcodes represented in the `zarr`do not represent genuine cells according to `empty_drops_result.csv`  (in one example bundle, only ~1% of rows are suggested to be real cells). To enable users to perform their own filtering on a reasonable set of barcodes while not delivering an abundance of low-quality data, a barcode is included in the index if the UMI count reported in `empty_drops_result.csv` is greater than 100, as entires with less are almost certainly artifactual. Additionally, users may still interact with these parameters via the `cell` table fields `barcode`, `total_umis` and `empty_drops_iscell`. | Within the index, cells are identified by a unique `cellkey` deterministically generated and managed by the Matrix Service. The current algorithm for generating an Optimus `cellkey` is the MD5 hash of the `bundle_uuid` and `barcode`. To ensure uniqueness and deterministic generation, the following assumptions must be true:<br /><br />1. A `bundle_uuid` must be preserved across updates and revisions to the bundle.<br />2. Barcodes are unique within a bundle.<br /><br />It should be noted that the event of a project reingestion does not maintain assumption (1) since bundle UUIDs are regenerated which results in generating a different `cellkey` for the same cell. The current response to this is to rebuild the index for the reingested project. |
+
+#### Metadata
+
+*Related tables:* `analysis`, `donor`, `specimen`, `cell_suspension`, `library_preparation`, `project`, `publication`, `contributor`
+
+The majority of metadata is parsed directly from metadata JSON files available in primary bundles. These fields are identifiable as fields with a valid _Canonical Metadata Names_ mapping in _Table 2.1_. The exceptions are "derived fields" such as `derived_organ_*` that require a degree of intellectual reasoning of the experiment to determine an accurate value. For example, the value of `derived_organ_ontology` may differ from the value of `organ_ontology`, which is read directly from `specimen_from_organism.organ.ontology`, as it is determined by the presence of organoids and/or cell lines in the experiment.
```

Comment:
Do you parse secondary analysis metadata as well?

Reply:
Ah, you're right we do. Good catch!

@justincc From our [earlier conversation](https://humancellatlas.slack.com/archives/C6SGU0R3J/p1574188428221000), you mentioned only `document_id`s  from `analysis_process_0.json` will be preserved on analysis bundle updates, correct? Our index stores `analysis_protocol`s and uses its `document_id` as a primary key. It sounds like the current AUDR analysis updates plan will in fact violate our assumptions on DIDs.

My question is, is it possible for these DIDs to be generated deterministically?

Reply:
The current plan is that submitters (Analysis) will be able to specify the UUID. So it would be up to them to generate a deterministic protocol UUID as well as one for the analysis process. @samanehsan and @rexwangcc could comment further.